pkg/stanza/fileconsumer/reader_factory.go

```diff
@@ -31,11 +33,26 @@ type readerFactory struct {
 	encodingConfig  helper.EncodingConfig
 }
 
-func (f *readerFactory) newReader(file *os.File, fp *Fingerprint) (*Reader, error) {
-	return f.newReaderBuilder().
+func (f *readerFactory) newReader(file *os.File, fp *Fingerprint, persister operator.Persister, hc *HeaderConfig) (*Reader, error) {
+	rb := f.newReaderBuilder().
 		withFile(file).
-		withFingerprint(fp).
-		build()
+		withFingerprint(fp)
+
+	if hc != nil {
+		enc, err := f.encodingConfig.Build()
+		if err != nil {
+			return nil, fmt.Errorf("failed to build encoding: %w", err)
+		}
+
+		h, err := hc.buildHeader(f.SugaredLogger, enc.Encoding, persister)
```

Comment:
Each reader needs to keep track of its own attributes, since each one represents a different file and potentially a different set of header attributes.

Reply:
It makes sense to me that we would need each reader to maintain a small amount of data representing the header text or fields, but all the rest of it seems unnecessary on a per-reader basis.

Reply:
I see what you mean, I'll see how I can refactor.

Reply:
I tried to pare this down a bit, so that we only compile the regex and splitter once.

The main state that we need to keep for each header are:
1. The attributes for the header
2. Whether we have actually read through the header fully or not ("finalized")
3. The current offset into the header (we need a separate offset since the reader may start at the end of the file, while the header should always be at the start)
4. The header pipeline, and its output

There's also the logger and config attached, which are both shared between many headers, and realistically we could pass these into the ReadHeader function, but I've left them on the struct to keep the function signature smaller.

Reply:
Thanks for taking another pass at it.

> the reader may start at the end of the file, while the header should always be at the start

Should we disallow this? Seems unlikely a user would want to start from the end of files if they also want the header.

Reply:
I think it's a useful mode of operation to read at end while also reading in the header data.

I'm thinking of the header data being useful metadata that applies to every log line, so it would be useful even if you only want only the new lines. E.g If I were reading a w3c formatted log, I'd still want to have the header metadata (which defines the fields to parse), even if I only want the most recent lines.

Reply:
The thing about the `start_at` setting is that it only applies to the first poll cycle. After that, any new files will always be read from the beginning anyways. So I think it comes down to whether or not users need this "read the header and then jump to the end" capability for such a limited time. The tradeoff between complexity and value seems questionable to me, but I can imagine we'd see this requested eventually so I'm ok either way.

Reply:
I'll leave it as-is for now, just because I'm sure it'll be something users of this feature will run into a lot, since `start_at` defaults to "end".